try_signal.cpp

```diff
@@ -219,38 +219,15 @@ void copy(iovec const* iov, std::size_t num_vecs)
 #else
 // windows
 
-namespace {
-
-sig::errors::error_code_enum map_exception_code(DWORD);
-
-void se_translator(unsigned int, _EXCEPTION_POINTERS* info)
-{
-	throw std::system_error(sig::map_exception_code(info->ExceptionRecord->ExceptionCode));
-}
-
-struct scoped_se_translator
-{
-	scoped_se_translator()
-	{ _prev_fun = _set_se_translator(se_translator); }
-
-	~scoped_se_translator()
-	{ _set_se_translator(_prev_fun); }
-
-	scoped_se_translator(scoped_se_translator const&) = delete;
-	scoped_se_translator& operator=(scoped_se_translator const&) = delete;
-
-private:
-	void (*_prev_fun)(unsigned int, struct _EXCEPTION_POINTERS*);
-};
-
-} // anonymous namespace
-
 void copy(iovec const* iov, std::size_t num_vecs)
 {
-	scoped_se_translator scope;
-
-	for (iovec const* end = iov + num_vecs; iov != end; ++iov)
-		std::memcpy(iov->dest, iov->src, iov->length);
+	__try {
+		for (iovec const* end = iov + num_vecs; iov != end; ++iov)
+			std::memcpy(iov->dest, iov->src, iov->length);
+	}
+	__except (EXCEPTION_EXECUTE_HANDLER) {
```

Comment:
This should only catch `EXCEPTION_IN_PAGE_ERROR`. Anything else is unrelated to I/O errors in the mmapped region, and probably fatal.

Reply:
I'm having second thoughts. In one way, it's nice to *just* get the I/O errors and correctly drop into the debugger on any other page error (which most likely indicate a bug, rather than a runtime error).

On the other hand, it may be a useful slightly more general facility to copy memory and turn page errors into runtime errors. I can't really think of a legitimate use case where it would be OK for the program to be uncertain about whether memory accesses are allowed or not though. The closest thing I can think of is collecting information in a crash handler to produce a report or a minidump.

It's probably better to keep it simple for now.

Reply:
however, testing this becomes significantly harder if it can only catch errors from memory mapped files. So for now I think I'll make it catch all memory errors (but at least exclude arithmetic and single-step exceptions)